File: node/src/ipc.ts
/**
 * Assembler IPC client — speaks the wire protocol from `hx_labs/crates/haap-ipc`.
 *
 * Wire format (verified against `crates/haap-ipc/src/framing.rs`):
 *
 *     [msg_len: u32 BE][msg_type: u8][payload: msg_len-1 bytes]
 *
 * `msg_len` includes the `msg_type` byte. `MAX_MESSAGE_SIZE` is 64 KiB.
 *
 * On connect both sides exchange an `IpcHandshake` (msg_type `0x00`) with
 * payload `[protocol_version: u16 BE][major: u16 BE][minor: u16 BE]
 * [patch: u16 BE][role: u8]`. Per `crates/haap-ipc/src/handshake.rs`, major
 * version MUST match; minor mismatches are logged warnings on the Rust side.
 *
 * Agent ↔ Assembler messages (CS v6.0.0 §39.7 channel allowlists):
 * - `MSG_TOOL_CALL_REQUEST = 0x52`    Agent → Assembler, JSON
 * - `MSG_TOOL_CALL_RESPONSE = 0x53`   Assembler → Agent, JSON
 * - `MSG_REQUEST_REJECTED = 0x54`     Assembler → Agent, JSON
 * - `MSG_CLARIFICATION_ANSWER = 0x61` Agent → Assembler, JSON (Profile E)
 */

import { Buffer } from "node:buffer";
import * as net from "node:net";

import { HandshakeError, IpcError, RequestRejected } from "./errors";

// ── Protocol constants (mirror crates/haap-ipc/src/handshake.rs) ────

export const PROTOCOL_VERSION = 1;
export const SDK_VERSION_MAJOR = 0;
export const SDK_VERSION_MINOR = 5;
export const SDK_VERSION_PATCH = 0;

export const ROLE_AGENT = 0x04;
export const ROLE_ASSEMBLER = 0x05;

export const MSG_TYPE_HANDSHAKE = 0x00;
export const MSG_TOOL_CALL_REQUEST = 0x52;
export const MSG_TOOL_CALL_RESPONSE = 0x53;
export const MSG_REQUEST_REJECTED = 0x54;
export const MSG_CLARIFICATION_ANSWER = 0x61;

export const MAX_MESSAGE_SIZE = 64 * 1024;

// ── Public wire types ───────────────────────────────────────────────

/**
 * Per-call outbound transport selector (CS v6.7.4 §34).
 *
 * `http_header` (default): token in `Authorization: HAAP <b64>` HTTP header.
 * `mcp_meta`: token in MCP `params._meta["haap/tbac"].token`.
 *
 * Values match serde `#[rename_all = "snake_case"]` on the Rust side.
 */
export enum TokenTransport {
  HttpHeader = "http_header",
  McpMeta = "mcp_meta",
}

/**
 * Agent → Assembler request (msg_type `0x52`).
 *
 * Mirrors `haap_ipc::messages::assembler::ToolCallRequest`. The Assembler
 * constructs the requested scope from `tool` / `action` / `resource` /
 * `constraints` per CS §39.7; the SDK process does not see token material
 * or session keys.
 */
export interface ToolCallRequest {
  requestId: string;
  targetRsUrl: string;
  httpMethod: string;
  headers?: Record<string, string>;
  tool: string;
  action?: readonly string[];
  resource?: string;
  constraints?: Record<string, unknown>;
  body?: Buffer | null;
  claimedIntentHash?: string;
  toolArguments?: unknown;
  contentType?: string;
  transport?: TokenTransport;
}

/**
 * Assembler → Agent response (msg_type `0x53`).
 *
 * Mirrors `haap_ipc::messages::assembler::ToolCallResponse`. `body` is the
 * decrypted RS response (`Buffer`).
 */
export interface ToolCallResponse {
  requestId: string;
  httpStatus: number;
  headers: Record<string, string>;
  body: Buffer;
}

function toolCallRequestToWire(req: ToolCallRequest): Record<string, unknown> {
  const wire: Record<string, unknown> = {
    request_id: req.requestId,
    target_rs_url: req.targetRsUrl,
    http_method: req.httpMethod,
    headers: req.headers ?? {},
    tool: req.tool,
    action: req.action ?? [],
    resource: req.resource ?? "*",
    constraints: req.constraints ?? {},
  };
  if (req.body !== undefined && req.body !== null) {
    wire.plaintext_request_body = req.body.toString("base64");
  }
  if (req.claimedIntentHash !== undefined) {
    wire.claimed_intent_hash = req.claimedIntentHash;
  }
  if (req.toolArguments !== undefined) {
    wire.tool_arguments = req.toolArguments;
  }
  if (req.contentType !== undefined) {
    wire.content_type = req.contentType;
  }
  if (req.transport !== undefined) {
    wire.transport = req.transport;
  }
  return wire;
}

function toolCallResponseFromWire(obj: Record<string, unknown>): ToolCallResponse {
  const requestId = String(obj.request_id ?? "");
  const httpStatus = Number(obj.http_status ?? 0);
  const headers = (obj.headers as Record<string, string>) ?? {};
  const bodyB64 = (obj.body as string) ?? "";
  return {
    requestId,
    httpStatus,
    headers,
    body: bodyB64 ? Buffer.from(bodyB64, "base64") : Buffer.alloc(0),
  };
}

// ── Framing helpers ─────────────────────────────────────────────────

export function encodeFrame(msgType: number, payload: Buffer): Buffer {
  const msgLen = 1 + payload.length;
  if (msgLen > MAX_MESSAGE_SIZE) {
    throw new IpcError(
      `frame too large: ${msgLen} bytes (max ${MAX_MESSAGE_SIZE})`,
    );
  }
  const frame = Buffer.allocUnsafe(4 + msgLen);
  frame.writeUInt32BE(msgLen, 0);
  frame.writeUInt8(msgType & 0xff, 4);
  payload.copy(frame, 5);
  return frame;
}

/**
 * Async reader that hands out [msgType, payload] tuples from a Node `net.Socket`.
 *
 * Internal state: a single resolver queue. The socket's `data` events accumulate
 * into a buffer; whenever a complete frame is present, the head waiter is
 * resolved with it. `error`, `end`, and `close` events propagate to all pending
 * waiters.
 */
class FrameReader {
  private buffer = Buffer.alloc(0);
  private waiters: Array<{
    resolve: (frame: [number, Buffer]) => void;
    reject: (err: Error) => void;
  }> = [];
  private closed = false;
  private closeError: Error | null = null;

  constructor(private readonly sock: net.Socket) {
    sock.on("data", (chunk: Buffer) => this.onData(chunk));
    sock.on("error", (err) => this.onError(err));
    sock.on("end", () => this.onClose(null));
    sock.on("close", () => this.onClose(null));
  }

  private onData(chunk: Buffer): void {
    this.buffer = Buffer.concat([this.buffer, chunk]);
    this.drain();
  }

  private onError(err: Error): void {
    this.onClose(err);
  }

  private onClose(err: Error | null): void {
    if (this.closed) return;
    this.closed = true;
    this.closeError =
      err ?? new IpcError("IPC peer closed connection mid-message");
    for (const w of this.waiters) {
      w.reject(this.closeError);
    }
    this.waiters = [];
  }

  private drain(): void {
    while (this.waiters.length > 0 && this.tryReadOne()) {
      // tryReadOne resolves head waiter
    }
  }

  private tryReadOne(): boolean {
    if (this.buffer.length < 4) return false;
    const msgLen = this.buffer.readUInt32BE(0);
    if (msgLen === 0) {
      const head = this.waiters.shift();
      head?.reject(new IpcError("frame length 0 (illegal)"));
      return true;
    }
    if (msgLen > MAX_MESSAGE_SIZE) {
      const head = this.waiters.shift();
      head?.reject(
        new IpcError(
          `frame too large: ${msgLen} bytes (max ${MAX_MESSAGE_SIZE})`,
        ),
      );
      return true;
    }
    if (this.buffer.length < 4 + msgLen) return false;
    const frame = this.buffer.subarray(4, 4 + msgLen);
    const msgType = frame.readUInt8(0);
    const payload = Buffer.from(frame.subarray(1));
    this.buffer = Buffer.from(this.buffer.subarray(4 + msgLen));
    const head = this.waiters.shift();
    head?.resolve([msgType, payload]);
    return true;
  }

  read(): Promise<[number, Buffer]> {
    if (this.closed) {
      return Promise.reject(
        this.closeError ?? new IpcError("connection closed"),
      );
    }
    return new Promise((resolve, reject) => {
      this.waiters.push({ resolve, reject });
      this.drain();
    });
  }
}

// ── Handshake ───────────────────────────────────────────────────────

function encodeHandshake(role: number): Buffer {
  const buf = Buffer.allocUnsafe(9);
  buf.writeUInt16BE(PROTOCOL_VERSION, 0);
  buf.writeUInt16BE(SDK_VERSION_MAJOR, 2);
  buf.writeUInt16BE(SDK_VERSION_MINOR, 4);
  buf.writeUInt16BE(SDK_VERSION_PATCH, 6);
  buf.writeUInt8(role & 0xff, 8);
  return buf;
}

interface DecodedHandshake {
  protocol: number;
  major: number;
  minor: number;
  patch: number;
  role: number;
}

function decodeHandshake(payload: Buffer): DecodedHandshake {
  if (payload.length < 9) {
    throw new IpcError(
      `handshake payload too short: ${payload.length} (want >=9)`,
    );
  }
  return {
    protocol: payload.readUInt16BE(0),
    major: payload.readUInt16BE(2),
    minor: payload.readUInt16BE(4),
    patch: payload.readUInt16BE(6),
    role: payload.readUInt8(8),
  };
}

// ── Connect + AssemblerClient ───────────────────────────────────────

/**
 * Connect to the Assembler at `endpoint` and return a Node `net.Socket`.
 *
 * On Unix, `endpoint` is a filesystem path to a Unix domain socket.
 * On Windows, `endpoint` is a Named Pipe path
 * (`\\\\.\\pipe\\haap-<agent_id>-agent-assembler-<index>`). Node's `net.connect`
 * supports both via the `path` option.
 */
export function connectAssembler(
  endpoint: string,
  timeoutMs: number,
): Promise<net.Socket> {
  return new Promise((resolve, reject) => {
    const sock = net.createConnection({ path: endpoint });
    const timer =
      timeoutMs > 0
        ? setTimeout(() => {
            sock.destroy();
            reject(new IpcError(`timed out connecting to ${endpoint}`));
          }, timeoutMs)
        : null;
    const onError = (err: Error) => {
      if (timer) clearTimeout(timer);
      reject(new IpcError(`connect ${endpoint} failed: ${err.message}`));
    };
    sock.once("error", onError);
    sock.once("connect", () => {
      if (timer) clearTimeout(timer);
      sock.removeListener("error", onError);
      resolve(sock);
    });
  });
}

/**
 * Synchronous-feeling, async client for the Assembler IPC channel.
 *
 * On `connect`, performs the version handshake (role = `Agent`). After that,
 * the connection is ready for ToolCallRequest / ToolCallResponse round-trips.
 *
 * Not safe to invoke concurrently. Wrap in a queue for parallel calls.
 */
export class AssemblerClient {
  private constructor(
    private readonly sock: net.Socket,
    private readonly reader: FrameReader,
  ) {}

  static async connect(
    endpoint: string,
    options: { timeoutMs?: number } = {},
  ): Promise<AssemblerClient> {
    const timeoutMs = options.timeoutMs ?? 5000;
    const sock = await connectAssembler(endpoint, timeoutMs);
    const reader = new FrameReader(sock);
    try {
      await write(sock, encodeFrame(MSG_TYPE_HANDSHAKE, encodeHandshake(ROLE_AGENT)));
      const [type, payload] = await reader.read();
      if (type !== MSG_TYPE_HANDSHAKE) {
        throw new IpcError(
          `expected handshake (0x00), got 0x${type.toString(16).padStart(2, "0")}`,
        );
      }
      const hs = decodeHandshake(payload);
      if (hs.major !== SDK_VERSION_MAJOR) {
        throw new HandshakeError(SDK_VERSION_MAJOR, hs.major);
      }
      if (hs.role !== ROLE_ASSEMBLER) {
        throw new IpcError(
          `expected peer role Assembler (0x05), got 0x${hs.role.toString(16).padStart(2, "0")}`,
        );
      }
    } catch (err) {
      sock.destroy();
      throw err;
    }
    return new AssemblerClient(sock, reader);
  }

  async invoke(req: ToolCallRequest): Promise<ToolCallResponse> {
    const wire = toolCallRequestToWire(req);
    const payload = Buffer.from(JSON.stringify(wire), "utf-8");
    await write(this.sock, encodeFrame(MSG_TOOL_CALL_REQUEST, payload));

    const [type, body] = await this.reader.read();
    if (type === MSG_TOOL_CALL_RESPONSE) {
      const obj = JSON.parse(body.toString("utf-8")) as Record<string, unknown>;
      return toolCallResponseFromWire(obj);
    }
    if (type === MSG_REQUEST_REJECTED) {
      const obj = JSON.parse(body.toString("utf-8")) as Record<string, unknown>;
      throw new RequestRejected(
        String(obj.request_id ?? req.requestId),
        String(obj.reason ?? ""),
      );
    }
    throw new IpcError(
      `unexpected response msg_type 0x${type.toString(16).padStart(2, "0")}; expected 0x53 or 0x54`,
    );
  }

  async sendClarificationAnswer(args: {
    pendingId: string;
    sessionId: number | bigint;
    answerIndex?: number;
    answerText?: string;
  }): Promise<void> {
    const obj: Record<string, unknown> = {
      pending_id: args.pendingId,
      session_id: typeof args.sessionId === "bigint" ? Number(args.sessionId) : args.sessionId,
    };
    if (args.answerIndex !== undefined) obj.answer_index = args.answerIndex;
    if (args.answerText !== undefined) obj.answer_text = args.answerText;
    const payload = Buffer.from(JSON.stringify(obj), "utf-8");
    await write(this.sock, encodeFrame(MSG_CLARIFICATION_ANSWER, payload));
  }

  close(): void {
    this.sock.destroy();
  }
}

function write(sock: net.Socket, data: Buffer): Promise<void> {
  return new Promise((resolve, reject) => {
    sock.write(data, (err) => {
      if (err) reject(new IpcError(`write failed: ${err.message}`));
      else resolve();
    });
  });
}

File: python/src/hawcx_haap/ipc.py
"""Assembler IPC client — speaks the wire protocol from ``hx_labs/crates/haap-ipc``.

Wire format (verified against ``crates/haap-ipc/src/framing.rs``):

    [msg_len: u32 BE][msg_type: u8][payload: msg_len-1 bytes]

``msg_len`` includes the ``msg_type`` byte (so on-wire bytes = 4 + msg_len).
``MAX_MESSAGE_SIZE`` is 64 KiB.

On connect, both sides exchange an ``IpcHandshake`` (msg_type 0x00) with payload
``[protocol_version: u16 BE][major: u16 BE][minor: u16 BE][patch: u16 BE][role: u8]``.
Per ``crates/haap-ipc/src/handshake.rs``, major version MUST match; minor
mismatches are logged warnings only.

Message types (CS v6.0.0 §39.7 channel allowlists, Agent ↔ Assembler):

- ``MSG_TOOL_CALL_REQUEST = 0x52``   — Agent → Assembler, JSON
- ``MSG_TOOL_CALL_RESPONSE = 0x53``  — Assembler → Agent, JSON
- ``MSG_REQUEST_REJECTED = 0x54``    — Assembler → Agent, JSON
- ``MSG_CLARIFICATION_ANSWER = 0x61``— Agent → Assembler, JSON (Profile E)

JSON schemas mirror the serde derives in
``crates/haap-ipc/src/messages/assembler.rs``.
"""

from __future__ import annotations

import base64
import json
import socket
import struct
import sys
from dataclasses import dataclass, field
from enum import Enum
from typing import Any

from hawcx_haap.errors import HandshakeError, IpcError, RequestRejected

# ── Protocol constants (mirror crates/haap-ipc/src/handshake.rs) ─────

PROTOCOL_VERSION: int = 1
SDK_VERSION_MAJOR: int = 0
SDK_VERSION_MINOR: int = 5
SDK_VERSION_PATCH: int = 0

ROLE_AGENT: int = 0x04
ROLE_ASSEMBLER: int = 0x05

MSG_TYPE_HANDSHAKE: int = 0x00
MSG_TOOL_CALL_REQUEST: int = 0x52
MSG_TOOL_CALL_RESPONSE: int = 0x53
MSG_REQUEST_REJECTED: int = 0x54
MSG_CLARIFICATION_ANSWER: int = 0x61

MAX_MESSAGE_SIZE: int = 64 * 1024


# ── Public types ────────────────────────────────────────────────────


class TokenTransport(str, Enum):
    """Per-call outbound transport selector (CS v6.7.4 §34).

    ``http_header`` (default): token in ``Authorization: HAAP <b64>`` header.
    ``mcp_meta``: token in MCP ``params._meta["haap/tbac"].token``.
    """

    HTTP_HEADER = "http_header"
    MCP_META = "mcp_meta"


@dataclass
class ToolCallRequest:
    """Agent → Assembler request (msg_type 0x52).

    Mirrors ``haap_ipc::messages::assembler::ToolCallRequest``. The Assembler
    constructs the requested scope from ``tool``/``action``/``resource``/
    ``constraints`` per CS §39.7; the Python process does not see token
    material or session keys.
    """

    request_id: str
    target_rs_url: str
    http_method: str
    headers: dict[str, str] = field(default_factory=dict)
    tool: str = ""
    action: list[str] = field(default_factory=list)
    resource: str = "*"
    constraints: dict[str, Any] = field(default_factory=dict)
    plaintext_request_body: bytes | None = None
    claimed_intent_hash: str | None = None
    tool_arguments: Any = None
    content_type: str | None = None
    transport: TokenTransport | None = None

    def to_wire(self) -> dict[str, Any]:
        out: dict[str, Any] = {
            "request_id": self.request_id,
            "target_rs_url": self.target_rs_url,
            "http_method": self.http_method,
            "headers": self.headers,
            "tool": self.tool,
            "action": self.action,
            "resource": self.resource,
            "constraints": self.constraints,
        }
        if self.plaintext_request_body is not None:
            out["plaintext_request_body"] = base64.b64encode(
                self.plaintext_request_body
            ).decode("ascii")
        if self.claimed_intent_hash is not None:
            out["claimed_intent_hash"] = self.claimed_intent_hash
        if self.tool_arguments is not None:
            out["tool_arguments"] = self.tool_arguments
        if self.content_type is not None:
            out["content_type"] = self.content_type
        if self.transport is not None:
            out["transport"] = self.transport.value
        return out


@dataclass
class ToolCallResponse:
    """Assembler → Agent response (msg_type 0x53).

    Mirrors ``haap_ipc::messages::assembler::ToolCallResponse``.
    """

    request_id: str
    http_status: int
    headers: dict[str, str]
    body: bytes

    @classmethod
    def from_wire(cls, obj: dict[str, Any]) -> ToolCallResponse:
        body_b64 = obj.get("body", "")
        body = base64.b64decode(body_b64) if body_b64 else b""
        return cls(
            request_id=obj["request_id"],
            http_status=int(obj["http_status"]),
            headers=dict(obj.get("headers") or {}),
            body=body,
        )


# ── Framing helpers (binary, used for handshake) ─────────────────────


def encode_frame(msg_type: int, payload: bytes) -> bytes:
    msg_len = 1 + len(payload)
    if msg_len > MAX_MESSAGE_SIZE:
        raise IpcError(
            f"frame too large: {msg_len} bytes (max {MAX_MESSAGE_SIZE})"
        )
    return struct.pack(">I", msg_len) + bytes([msg_type & 0xFF]) + payload


def _recv_exact(sock: socket.socket, n: int) -> bytes:
    buf = bytearray()
    while len(buf) < n:
        chunk = sock.recv(n - len(buf))
        if not chunk:
            raise IpcError("IPC peer closed connection mid-message")
        buf.extend(chunk)
    return bytes(buf)


def read_frame(sock: socket.socket) -> tuple[int, bytes]:
    """Read one ``[len: u32 BE][type: u8][payload]`` frame."""
    length_bytes = _recv_exact(sock, 4)
    msg_len = struct.unpack(">I", length_bytes)[0]
    if msg_len == 0:
        raise IpcError("frame length 0 (illegal)")
    if msg_len > MAX_MESSAGE_SIZE:
        raise IpcError(
            f"frame too large: {msg_len} bytes (max {MAX_MESSAGE_SIZE})"
        )
    body = _recv_exact(sock, msg_len)
    msg_type = body[0]
    payload = bytes(body[1:])
    return msg_type, payload


def write_frame(sock: socket.socket, msg_type: int, payload: bytes) -> None:
    sock.sendall(encode_frame(msg_type, payload))


# ── Handshake (mirrors crates/haap-ipc/src/handshake.rs) ─────────────


def _encode_handshake(role: int) -> bytes:
    return struct.pack(
        ">HHHHB",
        PROTOCOL_VERSION,
        SDK_VERSION_MAJOR,
        SDK_VERSION_MINOR,
        SDK_VERSION_PATCH,
        role & 0xFF,
    )


def _decode_handshake(payload: bytes) -> tuple[int, int, int, int, int]:
    if len(payload) < 9:
        raise IpcError(f"handshake payload too short: {len(payload)} (want >=9)")
    proto, major, minor, patch, role = struct.unpack(">HHHHB", payload[:9])
    return proto, major, minor, patch, role


def perform_handshake(sock: socket.socket, local_role: int = ROLE_AGENT) -> int:
    """Send local handshake, read peer handshake, validate major version.

    Returns the peer's role byte. Raises :class:`HandshakeError` on major
    version mismatch.
    """
    write_frame(sock, MSG_TYPE_HANDSHAKE, _encode_handshake(local_role))
    msg_type, payload = read_frame(sock)
    if msg_type != MSG_TYPE_HANDSHAKE:
        raise IpcError(
            f"expected handshake (0x00), got 0x{msg_type:02x}"
        )
    _proto, major, _minor, _patch, role = _decode_handshake(payload)
    if major != SDK_VERSION_MAJOR:
        raise HandshakeError(local_major=SDK_VERSION_MAJOR, remote_major=major)
    return role


# ── Platform-aware socket connect ────────────────────────────────────


def _connect_unix(path: str, timeout_secs: float | None) -> socket.socket:
    sock = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    if timeout_secs is not None:
        sock.settimeout(timeout_secs)
    sock.connect(path)
    return sock


def _connect_named_pipe(path: str, timeout_secs: float | None) -> _WindowsPipeSocket:
    # Lazy import — Windows-only path. Implemented in pipe_win.py.
    from hawcx_haap import pipe_win  # noqa: WPS433

    return pipe_win.connect(path, timeout_secs=timeout_secs)


def connect_assembler(endpoint: str, *, timeout_secs: float | None = 5.0) -> socket.socket:
    """Open a transport to the Assembler endpoint.

    On Unix, ``endpoint`` is a filesystem path to a UDS. On Windows, it is a
    Named Pipe path (``\\\\.\\pipe\\haap-<agent_id>-agent-assembler-<index>``).
    Returns a ``socket.socket``-compatible object (real ``socket`` on Unix; a
    file-handle-backed wrapper on Windows).
    """
    if sys.platform == "win32":
        return _connect_named_pipe(endpoint, timeout_secs)  # type: ignore[return-value]
    return _connect_unix(endpoint, timeout_secs)


# ── AssemblerClient ─────────────────────────────────────────────────


class AssemblerClient:
    """Synchronous client for the Assembler IPC channel.

    On construction, performs the version handshake (role = Agent). After that
    the connection is ready for ToolCallRequest / ToolCallResponse round-trips.
    """

    def __init__(self, sock: socket.socket) -> None:
        self._sock = sock

    @classmethod
    def connect(
        cls,
        endpoint: str,
        *,
        timeout_secs: float | None = 5.0,
    ) -> AssemblerClient:
        sock = connect_assembler(endpoint, timeout_secs=timeout_secs)
        try:
            peer_role = perform_handshake(sock, local_role=ROLE_AGENT)
        except Exception:
            try:
                sock.close()
            except Exception:
                pass
            raise
        if peer_role != ROLE_ASSEMBLER:
            try:
                sock.close()
            except Exception:
                pass
            raise IpcError(
                f"expected peer role Assembler (0x05), got 0x{peer_role:02x}"
            )
        return cls(sock)

    def invoke(self, req: ToolCallRequest) -> ToolCallResponse:
        """Send a ToolCallRequest; await ToolCallResponse or RequestRejected.

        Raises :class:`RequestRejected` if the Assembler rejects.
        Raises :class:`IpcError` on framing / transport errors.
        """
        payload = json.dumps(req.to_wire(), separators=(",", ":")).encode("utf-8")
        write_frame(self._sock, MSG_TOOL_CALL_REQUEST, payload)

        msg_type, body = read_frame(self._sock)
        if msg_type == MSG_TOOL_CALL_RESPONSE:
            obj = json.loads(body.decode("utf-8"))
            return ToolCallResponse.from_wire(obj)
        if msg_type == MSG_REQUEST_REJECTED:
            obj = json.loads(body.decode("utf-8"))
            raise RequestRejected(
                request_id=obj.get("request_id", req.request_id),
                reason=obj.get("reason", ""),
            )
        raise IpcError(
            f"unexpected response msg_type 0x{msg_type:02x}; "
            "expected 0x53 (ToolCallResponse) or 0x54 (RequestRejected)"
        )

    def send_clarification_answer(
        self,
        pending_id: str,
        session_id: int,
        *,
        answer_index: int | None = None,
        answer_text: str | None = None,
    ) -> None:
        """Profile E first hop: send a clarification answer (msg_type 0x61).

        Per CS v6.7.4 §39.7 the answer is forwarded by the Assembler to the
        TQS as the second hop (0x5E).
        """
        obj: dict[str, Any] = {
            "pending_id": pending_id,
            "session_id": int(session_id),
        }
        if answer_index is not None:
            obj["answer_index"] = int(answer_index)
        if answer_text is not None:
            obj["answer_text"] = answer_text
        payload = json.dumps(obj, separators=(",", ":")).encode("utf-8")
        write_frame(self._sock, MSG_CLARIFICATION_ANSWER, payload)

    def close(self) -> None:
        try:
            self._sock.close()
        except Exception:
            pass

    def __enter__(self) -> AssemblerClient:
        return self

    def __exit__(self, *_: Any) -> None:
        self.close()


# Forward declaration helper for type hints when pipe_win is missing on Unix.
class _WindowsPipeSocket:  # pragma: no cover — Windows-only
    """Shape placeholder; real impl in ``hawcx_haap.pipe_win``."""
